### PR TITLE
Fix printing to Avahi discovered "driverless" printers

### PR DESCRIPTION
**Pull request by NJRoadfan**
Previously trying to print to devices discovered by CUPS over Avahi/mDNS would fail unless cups-browsed is running (which auto-creates printer queues for all discovered driverless printers). Existing code assumed that all printer destinations reported by CUPS had a queue configured on the server. This PR instructs papd to check and connect to the selected printer destination, not just the CUPS server itself. This triggers CUPS to create a temporary printer queue in order to service print jobs.

### CODE REVIEW

**Comment by NJRoadfan** [comment]
To test that this PR with your "driverless" printer.

Disable cups-browsed if installed: `sudo systemctl stop cups-browsed`

Check that there are no configured printer queues on your local CUPS server. Your printer should appear after running `lpstat -e` but not `lpstat -v`

This change may be needed with CUPS 3.0 although I haven't confirmed. There is also a chance that a user may not have cups-browsed installed/enabled on their netatalk box even though most distros install it with CUPS by default.

**Comment by NJRoadfan** [comment]
Did some testing with PAP clients. LaserWriter 8.x, LaserWriter 7.x, and the GS/OS LaserWriter drivers all seem to be working without issue with this PR in place.

File: etc/papd/print_cups.c [hunk 1]
```diff
@@ -148,8 +167,6 @@ cups_printername_ok(char *name)         /* I - Name of printer */
 
         language = cupsLangDefault();
 
-	language = cupsLangDefault();
-
```
Comment (rdmark) [inline etc/papd/print_cups.c:152]: Thanks for fixing this! It seems I messed up with a merge conflict resolution last year.

**Comment by rdmark** [comment]
Good job narrowing down the cause and the fix. I wouldn't have guessed!
In fact, I think I ran into this very issue the other day when testing that fix in upstream netatalk2.
I had to jump through hoops deleting and recreating the printer queues, and restarting cups-browsed to get it working again.
This was the same old Raspbian Bullseye. I wonder what changed? 

Anyhow it seems the upcoming Debian Bookworm is still using cups 2.4.2, and CentOS Stream 2.3.3, so cups 3 adoption mustn't have taken off yet in the Linux world.

BTW what reference documentation do you use for the cups APIs?

**Comment by NJRoadfan** [comment]
CUPS API reference is here: https://openprinting.github.io/cups/doc/cupspm.html

CUPS 3.0 isn't planned to be released until late 2023 or early 2024. It'll be awhile until Debian switches, but I'm pretty sure Ubuntu will be pushing it out fairly quickly. There is ALOT of teeth gnashing over the dropping of printer drivers and PPDs and going to what basically amount to a IPP printer queuing system. You can find plenty online.

The behavior of temporary queues with driverless printers is poorly documented on both the end user side and the programming side. Some research revealed that GTK had problems with these queues for the same reason papd did. Printer dialogs would enumerate printer destinations, but not properly connect to them. Compounding this problem was cups-browsed, whose sole purpose is to "helpfully" detect and create permanent queues for driverless printers, thus hiding temporary queues (and bad programming) from the end user.

Next was the total lack of documentation in the CUPS API about how and when temporary queues are created. I landed up using one of the test programs that enumerates and queries printer features to determine exactly what triggers temporary queue creation (its cupsConnectDest() ). After that, it was just a matter of adding code to the correct places.

I will note that SAMBA's CUPS integration doesn't detect driverless printers at all. They appear to be using a really old way to querying for available printers which only picks up permanently configured queues. CUPS 3.0 will likely break it.

I plan on working on removing the calls to the CUPS PPD API shortly. Once that is replaced, it might be time to move these patches upstream for release. Plus we can finally remove all those lines about ignoring deprecation warnings.

**Comment by rdmark** [comment]
You are truly becoming a CUPS expert. When I did research to solve my own cups-browsed problem I did see a lot of complaints about printing woes in Gnome, and laments about how Apple had ruined CUPS (dunno if it's deserved or not.)

And yes I was going to ask about porting this to upstream. Good idea to make a batch of all your planned changes.

**Pull request merged**